### PR TITLE
[Data] Make `test_insert_xxx_optimization_rules` less brittle

### PR DESCRIPTION
**Pull request by bveeramani**
## Why are these changes needed?

`test_insert_logical_optimization_rules` and `test_insert_physical_optimization_rules` break if you add a rule to the respective default list. This PR refactors the test so that it doesn't break in this scenario.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/logical/optimizers.py [hunk 1]
```diff
@@ -32,20 +32,36 @@
...
+
     if insert_index is None:
         _LOGICAL_RULES.append(cls)
     else:
         _LOGICAL_RULES.insert(insert_index, cls)
 
 
+@DeveloperAPI
+def get_logical_rules() -> List[Type[Rule]]:
+    return list(_LOGICAL_RULES)
```
Comment (scottjlee) [inline python/ray/data/_internal/logical/optimizers.py:46]: do we need the `list()` here? it's already a list right?

File: python/ray/data/_internal/logical/optimizers.py [hunk 2]
```diff
@@ -32,20 +32,36 @@
...
+
     if insert_index is None:
         _PHYSICAL_RULES.append(cls)
     else:
         _PHYSICAL_RULES.insert(insert_index, cls)
 
 
+@DeveloperAPI
+def get_physical_rules() -> List[Type[Rule]]:
+    return list(_PHYSICAL_RULES)
```
Comment (scottjlee) [inline python/ray/data/_internal/logical/optimizers.py:62]: ditto

**Review by scottjlee — APPROVED** [review]
thanks for the fix!

File: python/ray/data/_internal/logical/optimizers.py (hunk 1 above)
Comment (bveeramani) [reply python/ray/data/_internal/logical/optimizers.py:46]: It's to make a copy so that you can't edit the internal data structure.

**Pull request merged**